h2o-clustering/src/main/java/water/clustering/api/GracefulAsyncRunner.java

```diff
@@ -0,0 +1,40 @@
+package water.clustering.api;
+
+import fi.iki.elonen.NanoHTTPD;
+import org.apache.log4j.Logger;
+
+
+/**
+ * The REST API is only needed until a valid flatfile is received. Afterwards, the flatfile is handed over to
+ * H2O via embedded config. Therefore, as soon as the embedded config is constructed, the REST API is shut down
+ * and the objects destroyed. However, there might still be a thread running - sending HTTP response to the assist which
+ * sent the flatfile. The {@link fi.iki.elonen.NanoHTTPD.DefaultAsyncRunner} does not wait for existing connections
+ * to be properly terminated and shuts them down.
+ * 
+ * This implementation overrides that behavior and waits until all connections are closed before shutdown.
+ */
+public class GracefulAsyncRunner extends NanoHTTPD.DefaultAsyncRunner {
+    private static final Logger LOG = Logger.getLogger(GracefulAsyncRunner.class);
+
+    @Override
+    public void closeAll() {
+        while (getRunning().size() > 0) {
+            synchronized (this) {
+                try {
+                    this.wait();
+                } catch (InterruptedException e) {
+                    LOG.error(e.getMessage(), e);
```

Comment:
do you really want to ignore interrupts? I would think not

Reply:
I am effectively not ignoring it, the handling of the interrupt means to ignore the graceful shutdown period and do a shutdown immediately (resources are freed). Which is done by the `super.closeAll()` call below. This call will be reached anyway.

Reply:
Oh now I see, I am not respecting the interrupt. Fixed by closing all connections immediately and self-interrupting.